File: Src/Compilers/CSharp/Source/Emitter/Model/ModuleReference.cs
// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using Microsoft.CodeAnalysis.CSharp.Symbols;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Text;
using Roslyn.Utilities;
using System.Diagnostics;
using System.Reflection;
using System.Collections.Immutable;

namespace Microsoft.CodeAnalysis.CSharp.Emit
{
    internal sealed class ModuleReference : Microsoft.Cci.IModuleReference, Microsoft.Cci.IFileReference
    {
        private readonly PEModuleBuilder moduleBeingBuilt;
        private readonly ModuleSymbol underlyingModule;

        internal ModuleReference(PEModuleBuilder moduleBeingBuilt, ModuleSymbol underlyingModule)
        {
            Debug.Assert(moduleBeingBuilt != null);
            Debug.Assert((object)underlyingModule != null);

            this.moduleBeingBuilt = moduleBeingBuilt;
            this.underlyingModule = underlyingModule;
        }

        void Microsoft.Cci.IReference.Dispatch(Microsoft.Cci.MetadataVisitor visitor)
        {
            visitor.Visit((Microsoft.Cci.IModuleReference)this);
        }

        string Microsoft.Cci.INamedEntity.Name
        {
            get
            {
                return underlyingModule.MetadataName;
            }
        }

        bool Microsoft.Cci.IFileReference.HasMetadata
        {
            get
            {
                return true;
            }
        }

        string Microsoft.Cci.IFileReference.FileName
        {
            get
            {
                return underlyingModule.Name;
            }
        }

        ImmutableArray<byte> Microsoft.Cci.IFileReference.GetHashValue(AssemblyHashAlgorithm algorithmId)
        {
            return underlyingModule.GetHash(algorithmId);
        }

        Microsoft.Cci.IAssemblyReference Microsoft.Cci.IModuleReference.GetContainingAssembly(Microsoft.CodeAnalysis.Emit.Context context)
        {
            if (this.moduleBeingBuilt.OutputKind.IsNetModule() &&
                ReferenceEquals(moduleBeingBuilt.SourceModule.ContainingAssembly, underlyingModule.ContainingAssembly))
            {
                return null;
            }

            return moduleBeingBuilt.Translate(underlyingModule.ContainingAssembly, context.Diagnostics);
        }

        public override string ToString()
        {
            return underlyingModule.ToString();
        }

        IEnumerable<Microsoft.Cci.ICustomAttribute> Microsoft.Cci.IReference.GetAttributes(Microsoft.CodeAnalysis.Emit.Context context)
        {
            return SpecializedCollections.EmptyEnumerable<Microsoft.Cci.ICustomAttribute>();
        }

        Microsoft.Cci.IDefinition Microsoft.Cci.IReference.AsDefinition(Microsoft.CodeAnalysis.Emit.Context context)
        {
            return null;
        }
    }
}

File: Src/Compilers/VisualBasic/Source/Emit/ModuleReference.vb
' Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System
Imports System.Collections.Generic
Imports System.Collections.Immutable
Imports System.Reflection
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.Emit

    Friend NotInheritable Class ModuleReference
        Implements Microsoft.Cci.IModuleReference
        Implements Microsoft.Cci.IFileReference

        Private ReadOnly m_ModuleBeingBuilt As PEModuleBuilder
        Private ReadOnly m_UnderlyingModule As ModuleSymbol

        Friend Sub New(moduleBeingBuilt As PEModuleBuilder, underlyingModule As ModuleSymbol)
            Debug.Assert(moduleBeingBuilt IsNot Nothing)
            Debug.Assert(underlyingModule IsNot Nothing)
            Me.m_ModuleBeingBuilt = moduleBeingBuilt
            Me.m_UnderlyingModule = underlyingModule
        End Sub

        Private Sub IReferenceDispatch(visitor As Microsoft.Cci.MetadataVisitor) Implements Microsoft.Cci.IReference.Dispatch
            visitor.Visit(DirectCast(Me, Microsoft.Cci.IModuleReference))
        End Sub

        Private ReadOnly Property INamedEntityName As String Implements Microsoft.Cci.INamedEntity.Name
            Get
                Return m_UnderlyingModule.Name
            End Get
        End Property

        Private ReadOnly Property IFileReferenceHasMetadata As Boolean Implements Microsoft.Cci.IFileReference.HasMetadata
            Get
                Return True
            End Get
        End Property

        Private ReadOnly Property IFileReferenceFileName As String Implements Microsoft.Cci.IFileReference.FileName
            Get
                Return m_UnderlyingModule.Name
            End Get
        End Property

        Private Function IFileReferenceGetHashValue(algorithmId As AssemblyHashAlgorithm) As ImmutableArray(Of Byte) Implements Microsoft.Cci.IFileReference.GetHashValue
            Return m_UnderlyingModule.GetHash(algorithmId)
        End Function

        Private Function IModuleReferenceGetContainingAssembly(context As Microsoft.CodeAnalysis.Emit.Context) As Microsoft.Cci.IAssemblyReference Implements Microsoft.Cci.IModuleReference.GetContainingAssembly
            If m_ModuleBeingBuilt.OutputKind.IsNetModule() AndAlso
                m_ModuleBeingBuilt.SourceModule.ContainingAssembly Is m_UnderlyingModule.ContainingAssembly Then
                Return Nothing
            End If

            Return m_ModuleBeingBuilt.Translate(m_UnderlyingModule.ContainingAssembly, context.Diagnostics)
        End Function

        Public Overrides Function ToString() As String
            Return m_UnderlyingModule.ToString()
        End Function

        Private Function IReferenceAttributes(context As Microsoft.CodeAnalysis.Emit.Context) As IEnumerable(Of Microsoft.Cci.ICustomAttribute) Implements Microsoft.Cci.IReference.GetAttributes
            Return SpecializedCollections.EmptyEnumerable(Of Microsoft.Cci.ICustomAttribute)()
        End Function

        Private Function IReferenceAsDefinition(context As Microsoft.CodeAnalysis.Emit.Context) As Microsoft.Cci.IDefinition Implements Microsoft.Cci.IReference.AsDefinition
            Return Nothing
        End Function
    End Class
End Namespace
